### DISCUSSION TITLE
Promotion of data families

### DISCUSSION THREAD

**Discussion by int-index** [ghc-proposals/ghc-proposals#456] [Ideas]
At the moment, GHC does not support promotion of data families, as stated by the User’s Guide. See [6.4.10. Datatype promotion](https://downloads.haskell.org/ghc/9.2.1/docs/html/users_guide/exts/data_kinds.html):

> Data family instance constructors cannot be promoted at the moment. GHC’s type theory just isn’t up to the task of promoting data families, which requires full dependent types.

What exactly is missing to permit this? In [#15245](https://gitlab.haskell.org/ghc/ghc/-/issues/15245) @goldfirere says:

> we need to abstract over a coercion to promote data family instances properly

When does this arise? Could we promote some data families, but not all? In particular, here’s a data family that I’d like to promote:

```haskell
{-# LANGUAGE TypeFamilies, PolyKinds #-}

module PromoteDataFam where

import Data.Kind

data family Tuple :: k

data instance Tuple
  = Tup0

data instance Tuple (a :: Type)
  = Tup1 a

data instance Tuple (a :: Type) (b :: Type)
  = Tup2 a b

data instance Tuple (a :: Type) (b :: Type) (c :: Type)
  = Tup3 a b c
```

In GHC 9.2.1, asking for the kind of `Tup3` gives me this error:

```
ghci> :k Tup3

<interactive>:1:1: error:
    • Data constructor ‘Tup3’ cannot be used here
        (it comes from a data family instance)
    • In the type ‘Tup3’
```

My motivation comes from #281. To cite the proposal:

> For tuples of higher arities, users may want to define their own synonyms [...] We do not propose adding them to `Data.Tuple` for the time being, as better APIs are possible (e.g. based on variadic data families) but are blocked by other technical issues. We leave it as future work.

But it would be nicer to get this right from the get go. In particular, under `-XNoListTupleTypeSyntax` (part of the proposal), how do we _print_ tuples? What should be the output of

```
ghci> :set -XNoListTupleTypeSyntax
ghci> :t (True, length [])
```

Today the answer is `(Bool, Int)`. But that wouldn’t be the correct answer under `-XNoListTupleTypeSyntax`! I would rather print `Tuple Bool Int`.

And for a triple such as `(True, length [], 'x')` it could be `Tuple Bool Int Char`, and so on. If we could make the `Tuple` data family built-in, that would be possible. I’d also define it slightly differently to use built-in syntax:

```
data instance Tuple (a :: Type) (b :: Type)
  = (a, b)

data instance Tuple (a :: Type) (b :: Type) (c :: Type)
  = (a, b, c)
```

So the `Tup0` `Tup1` `Tup2` `Tup3` constructors above are just to demonstrate the issue of promotion. In the final design I intend to go with the normal `(a, b, c)` syntax.

----

Q: Why not use a type family?
A: Class instances. 

With a data family, we could define e.g.

```
instance Functor (Tuple a) where
  fmap f (a, b) = (a, f b)

instance Functor (Tuple a b) where
  fmap f (a, b, c) = (a, b, f c)
```

I’d want `Tuple` to be usable anywhere the tuple type constructor is usable today, so that the transition would be seamless.

----

Any thoughts on this, @goldfirere, @simonpj, @RyanGlScott, others? Could we support promotion of `Tuple` specifically if not all data families in general? And if so, could identify a set of restrictions that makes it possible?

**Comment by simonpj** [comment]
The big thing we lack in the type language is any way to do pattern matching that binds coercion variables; in Core this is just case expression.

However, data families don't need this.  When we have
```
data family T a

data instance T [b] = T1 b | T2
```
we behave very much like
```
data R:TList b = R:T1 b | R:T2_
-- R:T1_ :: b -> TList b

-- axTList :: forall b.  R:TList b ~ T [b]

T1 = /\b \(x::b). R:T1 @b x |>  axList @b

pattern T1 x <- ( (\t -> t |> sym axList) -> R:T1 x )
```
That is, we define a new, ordinary algebraic data type `R:TList`, with ordinary data constructors `R:T1` and `R:T2`.  (I'm not 100% sure of the naming, but the idea is to use names that the user can't use, with a colon in the middle.)

Then any *use* of `T1` in source code just inlines to a use of `R:T1` with a cast.

Similarly *pattern-matching* on `T1` in source code turns into pattern matching on `R:T1`, but first casting.

We'd need to do this in types too.   But we need to be careful never to expose these `R:` things to the user.   I have not thought through how to do this.  Maybe not too hard.

**Comment by int-index** [comment]
@goldfirere mentioned to me elsewhere that his comment about the need for coercion abstraction only applies if we take constraints into account. So it should be possible to promote simple data families.

**Comment by typedrat** [comment]
This would go deliciously with an expansion of #106 to allow kind families, giving us open kinds without resorting to typo-vulnerable abuse of `Symbol`.
